gnat2goto/driver/tree_walk.adb

```diff
@@ -694,31 +694,57 @@ package body Tree_Walk is
       Resolved_Underlying : constant Irep :=
         Follow_Symbol_Type (Underlying, Global_Symbol_Table);
       --  ??? why not get this from the entity
+
+      function Get_Bound_Expr (Bound_Node : Node_Id) return Bound_Type_Expr;
+      function Get_Bound_Expr (Bound_Node : Node_Id) return Bound_Type_Expr
+      is
+         Attribute_Not_Range_Bound : exception;
+      begin
+         case Get_Attribute_Id (Attribute_Name (Bound_Node)) is
+            when Attribute_First =>
+               return Bound_Type_Expr (Do_Array_First (Bound_Node));
+            when Attribute_Last  =>
+               return Bound_Type_Expr (Do_Array_Last (Bound_Node));
+               when others          => raise Attribute_Not_Range_Bound;
+         end case;
+      end Get_Bound_Expr;
+
+      Lower_Bound : constant Node_Id := Low_Bound (Range_Expr);
+      Upper_Bound : constant Node_Id := High_Bound (Range_Expr);
+
+      Lower_Bound_Value : Integer;
+      Upper_Bound_Value : Integer;
+
+      Unsupporter_Range_Kind : exception;
+
+      Result_Type : constant Irep := New_Irep (I_Bounded_Signedbv_Type);
    begin
       if not (Kind (Resolved_Underlying) in Class_Bitvector_Type) then
          Report_Unhandled_Node_Empty (Range_Expr, "Do_Base_Range_Constraint",
                                       "range expression not bitvector type");
-         return R : constant Irep := New_Irep (I_Bounded_Signedbv_Type);
+         return Result_Type;
       end if;
-      if Nkind (Low_Bound (Range_Expr)) /= N_Integer_Literal then
-         Report_Unhandled_Node_Empty (Range_Expr, "Do_Base_Range_Constraint",
-                                     "low bound range expression not literal");
-         return R : constant Irep := New_Irep (I_Bounded_Signedbv_Type);
-      end if;
-      if Nkind (High_Bound (Range_Expr)) /= N_Integer_Literal then
-         Report_Unhandled_Node_Empty (Range_Expr, "Do_Base_Range_Constraint",
-                                    "high bound range expression not literal");
-         return R : constant Irep := New_Irep (I_Bounded_Signedbv_Type);
-      end if;
-      return R : constant Irep := New_Irep (I_Bounded_Signedbv_Type) do
-         Set_Width (R, Get_Width (Resolved_Underlying));
-         Set_Lower_Bound (I     => R,
-                          Value => Store_Bound (Bound_Type (Intval (
-                            Low_Bound (Range_Expr)))));
-         Set_Upper_Bound (I     => R,
-                          Value => Store_Bound (Bound_Type (Intval (
-                            High_Bound (Range_Expr)))));
-      end return;
+
+      case Nkind (Lower_Bound) is
+         when N_Integer_Literal => Lower_Bound_Value :=
+            Store_Nat_Bound (Bound_Type_Nat (Intval (Lower_Bound)));
+         when N_Attribute_Reference => Lower_Bound_Value :=
+            Store_Expr_Bound (Get_Bound_Expr (Lower_Bound));
+         when others => raise Unsupporter_Range_Kind;
```

Comment:
Should be `assert`

Reply:
It can reasonably happen so I changed it to Unsupported feature.